### ISSUE TITLE
Understand SAC v1 replay buffer abnormality

### ISSUE THREAD

**Issue by ChengshuLi** [tensorflow/agents#192]
Hi, I have a quick question for the following code: SAC v1.
https://github.com/tensorflow/agents/blob/a614c5b699ab08f6ac501c1d2bfc520d0a46572f/tf_agents/agents/sac/examples/v1/train_eval.py#L222-L228
which is very different from v2 and also other `train_eval.py` code, say DDPG. 
https://github.com/tensorflow/agents/blob/a614c5b699ab08f6ac501c1d2bfc520d0a46572f/tf_agents/agents/sac/examples/v2/train_eval.py#L263-L266

Specifically, why do we need to 
1) use 5 * batch_size
2) do unbatch and batch
3) filter trajectory that is at the boundary of episodes (`_filter_invalid_transition`)

Thanks a lot for your help!

**Comment by samfishman** [comment]
Hi Chengshu,

Thanks for the question. Bear with me, because it is fairly complicated.

At a high level, the reason we do this is to filter transitions from the end of one episode to the start of the next. These transitions are artificially produced by the replay buffer as an artifact of the fact that we continually add time steps we receive from the environment and return them as they are in the buffer. It doesn't make sense to learn from these transitions because it doesn't represent anything that actually happens in the environment.

As for why we do this here and not in any other agents - other agents sometimes filter these invalid transitions by applying a mask on the loss (e.g. [DQN](https://github.com/tensorflow/agents/blob/master/tf_agents/agents/dqn/dqn_agent.py#L416-L417)). DDPG [has a TODO](https://github.com/tensorflow/agents/blob/master/tf_agents/agents/ddpg/ddpg_agent.py#L224) to do something like this. However, the masking method isn't great because the number of valid transitions will vary from batch to batch, resulting in varying effective batch size from train step to train step. It is usually important to keep a consistent train step so that each batch of gradients is treated equally.

The reason we unbatch and batch is that different elements of the batch will either be valid or not, so we need to filter them individually (hence the unbatch), but we then need to rebatch to send it to the agent's train method.

We use 5*batch_size as an efficiency measure - moving more data at once has higher throughput. However, we still want to use the specified batch_size (not multiplied by 5) for training since it can be very costly to calculate the loss and backprop on larger batch sizes and learning algorithms are often sensitive to batch size.

Presently, we are working on modifying our replay buffers to always return full batches of valid transitions (i.e. no boundary transitions) which means that agents won't need to worry about varying batch sizes and train_eval scripts won't need to filter anything out. At that point, all of our agents will be more aligned.

Thanks again for the question!

**Comment by ChengshuLi** [comment]
@samfishman

Thanks for your answer. It's super helpful.

One quick follow-up question. Based on [api docs](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#prefetch), `prefetch`'s argument `buffer_size` stands for how many batches to prefetch. Should the code below be `.prefetch(5)` instead of `.prefetch(5 * batch_size)`? 
https://github.com/tensorflow/agents/blob/a614c5b699ab08f6ac501c1d2bfc520d0a46572f/tf_agents/agents/sac/examples/v1/train_eval.py#L222-L228

**Comment by samfishman** [comment]
I think you're right about this. Feel free to submit a PR to fix it!
